library/choose_disk.py

```diff
@@ -353,46 +353,69 @@ def expand_disks(lookup_disks, ceph_type="", module=None):
     return final_disks
 
 
-def disk_label(partition):
+def disk_label(partition, current_fsid, ceph_disk):
     '''
     Reports if a partition is containing some ceph structures
     '''
-    # 1) let's search for legacy metadata made by ceph-disk
-    stdout = subprocess.check_output(["blkid", "-s", "PARTLABEL", "-o", "value", "{}".format(partition)])
 
+    # 1) let's search for metadata coming from ceph-volume
+    try:
+        output_cmd = subprocess.Popen(["ceph-volume", "lvm", "list", "--format=json", partition],
+                                      stdout=subprocess.PIPE)
+        raw_json, _ = output_cmd.communicate()
+        try:
+            json_dict = json.loads(raw_json)
+        except (ValueError, TypeError) as e:
+            fatal("Cannot parse ceph-volume properly : {}".format(e))
+
+        # If the disk have some metadata, the json is filled with information
+        for key, value in json_dict.items():
+            # As we parse a partition it should only have a single entry unless that's an error
+            if (len(value) > 1):
+                fatal("Error while parsing ceph-volume")
+            if "tags" in value[0]:
+                if "ceph.cluster_fsid" in value[0]['tags']:
+                    if value[0]["tags"]["ceph.cluster_fsid"] != current_fsid:
+                        return "foreign : {}".format(value[0]["tags"]["ceph.cluster_fsid"])
+            if "type" in value[0]:
+                return value[0]["type"]
+            else:
+                # ceph-volume should give a type
+                # If not, we don't really what it is _but_ the disks is used
+                # So we should ignore it, let's report it undefined
+                return "undefined"
+
+    except subprocess.CalledProcessError:
+        pass
+
+    # 2) let's try with ceph-disk
+    # If the disk have some metadata, the json is filled with information
+    for item in ceph_disk:
```

Comment:
the cyclomatic complexity for this block is too high, ideally you wouldn't want to nest these if/for/if/if/if/if in this way because it is hard to understand what is going on. Breaking up these conditionals will make it better.

Reply:
This code is running by successive exclusion. We might split the cases in various functions.